File: hello-aws-lambda-nodejs12.x/fnHello.js
const axios = require("axios")

const pino = require("pino")
const LOGGER = pino({ level: process.env.LOG_LEVEL || "info" })

const AWSXRay = require("aws-xray-sdk-core")
AWSXRay.captureHTTPsGlobal(require("http"));

const STATIC_RANDOM = Math.random()

exports.apiAsyncLambdaHandler = async (event, context) => {
    request = JSON.parse(event["body"])
    return {
        "statusCode": 200,
        "body": JSON.stringify(await internalAsyncLambdaHandler(request, context))
    }
}

exports.asyncLambdaHandler = internalAsyncLambdaHandler

async function internalAsyncLambdaHandler(event, context) {
    LOGGER.info(getRuntimeInfo())
    LOGGER.info(await getPublicIP())

    LOGGER.info(`${process.env.AWS_EXECUTION_ENV} called ${context.functionName} has ${context.getRemainingTimeInMillis() / 1000} seconds to live`)

    LOGGER.info(`Static value: ${STATIC_RANDOM}, invocation value: ${Math.random()}`)

    return processRequest(event, process.env.PACKAGE_MANAGEMENT_SYSTEM, "async")
}

exports.apiSyncLambdaHandler = function(event, context, callback) {
    request = JSON.parse(event["body"])

    LOGGER.info(getRuntimeInfo())

    LOGGER.info(`${process.env.AWS_EXECUTION_ENV} called ${context.functionName} has ${context.getRemainingTimeInMillis() / 1000} seconds to live`)

    LOGGER.info(`Static value: ${STATIC_RANDOM}, invocation value: ${Math.random()}`)

    callback(null, {
        "statusCode": 200,
        "body": JSON.stringify(processRequest(request, process.env.PACKAGE_MANAGEMENT_SYSTEM, "non-async"))
    })
}

exports.syncLambdaHandler = function(event, context, callback) {
    LOGGER.info(getRuntimeInfo())

    LOGGER.info(`${process.env.AWS_EXECUTION_ENV} called ${context.functionName} has ${context.getRemainingTimeInMillis() / 1000} seconds to live`)

    LOGGER.info(`Static value: ${STATIC_RANDOM}, invocation value: ${Math.random()}`)

    callback(null, processRequest(event, process.env.PACKAGE_MANAGEMENT_SYSTEM, "non-async"))
}

exports.exportedProcessRequest = (event, packageManagementSystem) => {
    return processRequest(event, packageManagementSystem)
}

function processRequest(event, packageManagementSystem, type) {
    if (event["throwError"]) {
        throw new Error("Sorry, but the caller wants to me to throw an error");
    }

    const greeting = `Hello ${event["firstName"]} ${event["lastName"]}, this is a Nodejs12.x AWS Lambda ${type} function developed using ${packageManagementSystem}!`

    const result = `Your age in hexadecimal is ${parseInt(event["age"], 10).toString(16)}`;

    return {
        "greeting": greeting,
        "result": result
    }
}

function getRuntimeInfo() {
    return AWSXRay.captureFunc("getRuntimeInfo", function(){
        return `node.js ${process.versions.node} v8 ${process.versions.v8}`
    });
}

async function getPublicIP() {
    try {
        const response = await axios("http://checkip.amazonaws.com/");
        if (response.status / 100 == 2) {
            return `Public IP:  ${response.data.trim()}`
        } else {
            return `Unknown public IP, status code ${response.status} received`
        }
    } catch (err) {
        LOGGER.error(err);
        return `Error: ${err.toString()}`
    }
}

File: hello-aws-lambda-python3.8/app/fn_hello.py
import logging
import platform
import os
import random
import requests

from aws_xray_sdk.core import xray_recorder
from aws_xray_sdk.core import patch_all

patch_all()

LOGGER = logging.getLogger()
LOGGER.setLevel(logging.INFO)

STATIC_RANDOM = random.random()


class SimpleException(Exception):
    pass


def lambda_handler(event, context):
    LOGGER.info(get_runtime_info())
    LOGGER.info(get_public_ip())

    LOGGER.info("{} called {} has {} seconds to live".format(
        os.environ["AWS_EXECUTION_ENV"], context.function_name, context.get_remaining_time_in_millis() / 1000))

    LOGGER.info("Static value : {}, invocation value: {}".format(STATIC_RANDOM, random.random()))

    return process_request(event, os.environ["PACKAGE_MANAGEMENT_SYSTEM"])


def process_request(event, package_management_system):
    if event["throw_error"]:
        raise SimpleException("Sorry, but the caller wants to me to throw an error")

    greeting = "Hello {} {}, this is a Python3.8 AWS Lambda function developed using {}!".format(
        event["first_name"], event["last_name"], package_management_system)

    result = "Your age in hexadecimal is {}".format(hex(int(event["age"])))

    return {
        "greeting": greeting,
        "result": result
    }


@xray_recorder.capture("getRuntimeInfo")
def get_runtime_info():
    return "{} {} {}".format(platform.python_implementation(), platform.python_version(), platform.python_compiler())


def get_public_ip():
    try:
        response = requests.get("http://checkip.amazonaws.com/")
        if response.status_code / 100 == 2:
            return "Public IP: {}".format(response.text.strip())
        else:
            return "Unknown public IP, status code {} received".format(response.status_code)
    except requests.RequestException as e:
        LOGGER.error(e)
        return "RequestException: {}".format(str(e))
